File: julia/AIInAction/src/ch201_batch_norm.jl
"""
    Ch201BatchNorm

Batch Normalization forward and backward pass from scratch (Julia).

Mirrors the Python module `aiinaction.ch201_batch_norm` and the Rust module
`aiinaction::ch201_batch_norm`. The shared fixtures in `test/test_ch201_batch_norm.jl`
match the Python/Rust suites, which keeps the three implementations at parity.

For a mini-batch `X` of shape `(m, d)`, the forward transform standardizes each
feature over the batch (population variance, `ddof = 0`) and applies a learnable
scale `gamma` and shift `beta`:

    mu_j    = mean_i X[i, j]
    var_j   = mean_i (X[i, j] - mu_j)^2
    xhat    = (X .- mu') ./ sqrt.(var .+ eps)'
    Y       = gamma' .* xhat .+ beta'
"""
module Ch201BatchNorm

export BatchNormCache, batch_norm_forward, batch_norm_backward, batch_norm_inference

"""Intermediate quantities saved by the forward pass for the backward pass."""
struct BatchNormCache
    x_hat::Matrix{Float64}
    inv_std::Vector{Float64}
    gamma::Vector{Float64}
    mean::Vector{Float64}
    var::Vector{Float64}
end

function _as_matrix(X, name="X")
    A = Matrix{Float64}(X)
    n, d = size(A)
    n >= 1 || throw(ArgumentError("$name must have at least one row"))
    d >= 1 || throw(ArgumentError("$name must have at least one feature"))
    all(isfinite, A) || throw(ArgumentError("$name contains non-finite values (nan or inf)"))
    return A
end

function _as_vector(v, d, name)
    a = Vector{Float64}(v)
    length(a) == d || throw(ArgumentError("$name has length $(length(a)) but X has $d features"))
    all(isfinite, a) || throw(ArgumentError("$name contains non-finite values (nan or inf)"))
    return a
end

"""
    batch_norm_forward(X, gamma, beta; eps=1e-5)

Training-time Batch Normalization forward pass. Returns `(Y, cache)` where `Y` is
the `m x d` output and `cache::BatchNormCache` holds the backward-pass state.
"""
function batch_norm_forward(X, gamma, beta; eps::Float64=1e-5)
    eps > 0 || throw(ArgumentError("eps must be positive, got $eps"))
    A = _as_matrix(X)
    n, d = size(A)
    g = _as_vector(gamma, d, "gamma")
    b = _as_vector(beta, d, "beta")

    mu = vec(sum(A; dims=1)) ./ n
    Xc = A .- mu'
    var = vec(sum(abs2, Xc; dims=1)) ./ n  # population variance (ddof=0)
    inv_std = 1.0 ./ sqrt.(var .+ eps)
    x_hat = Xc .* inv_std'
    Y = x_hat .* g' .+ b'

    return Y, BatchNormCache(x_hat, inv_std, g, mu, var)
end

"""
    batch_norm_backward(dY, cache)

Backward pass. Returns `(dX, dgamma, dbeta)` with `dX` of shape `(m, d)` and the
parameter gradients of length `d`.
"""
function batch_norm_backward(dY, cache::BatchNormCache)
    D = Matrix{Float64}(dY)
    size(D) == size(cache.x_hat) ||
        throw(ArgumentError("dY has shape $(size(D)) but cache was built for $(size(cache.x_hat))"))
    all(isfinite, D) || throw(ArgumentError("dY contains non-finite values (nan or inf)"))
    n = size(D, 1)

    dgamma = vec(sum(D .* cache.x_hat; dims=1))
    dbeta = vec(sum(D; dims=1))

    g = D .* cache.gamma'                       # dL/dxhat
    g_mean = vec(sum(g; dims=1)) ./ n
    gxhat_mean = vec(sum(g .* cache.x_hat; dims=1)) ./ n
    dX = cache.inv_std' .* (g .- g_mean' .- cache.x_hat .* gxhat_mean')

    return dX, dgamma, dbeta
end

"""
    batch_norm_inference(X, gamma, beta, running_mean, running_var; eps=1e-5)

Inference-time Batch Normalization using frozen population statistics. Applies
`y = gamma * (x - running_mean) / sqrt(running_var + eps) + beta` per feature.
"""
function batch_norm_inference(X, gamma, beta, running_mean, running_var; eps::Float64=1e-5)
    eps > 0 || throw(ArgumentError("eps must be positive, got $eps"))
    A = _as_matrix(X)
    n, d = size(A)
    g = _as_vector(gamma, d, "gamma")
    b = _as_vector(beta, d, "beta")
    rm = _as_vector(running_mean, d, "running_mean")
    rv = _as_vector(running_var, d, "running_var")
    any(rv .< 0) && throw(ArgumentError("running_var must be non-negative"))

    return g' .* (A .- rm') ./ sqrt.(rv .+ eps)' .+ b'
end

end # module Ch201BatchNorm

File: src/aiinaction/ch201_batch_norm.py
"""Batch Normalization forward and backward pass from scratch.

A small, well-validated reference implementation of Batch Normalization (Ioffe and
Szegedy, 2015). The public API mirrors the Julia (`AIInAction.Ch201BatchNorm`) and
Rust (`aiinaction::ch201_batch_norm`) implementations one-to-one; the
cross-language parity tests assert that all three agree to within floating-point
tolerance on shared fixtures.

For a mini-batch ``X`` of shape ``(m, d)`` (``m`` examples, ``d`` features), the
forward transform standardizes every feature over the batch and then applies a
learnable scale ``gamma`` and shift ``beta``:

.. math::

    \\mu_j = \\frac1m \\sum_i x_{ij}, \\qquad
    \\sigma_j^2 = \\frac1m \\sum_i (x_{ij} - \\mu_j)^2, \\\\
    \\hat{x}_{ij} = \\frac{x_{ij} - \\mu_j}{\\sqrt{\\sigma_j^2 + \\epsilon}}, \\qquad
    y_{ij} = \\gamma_j \\hat{x}_{ij} + \\beta_j .

The variance uses the population (biased, ``ddof=0``) convention, matching the
original paper and essentially every deep-learning framework's training-time BN.

The backward pass propagates an upstream gradient ``dY`` of shape ``(m, d)`` to
``dX``, ``dgamma`` and ``dbeta`` using the closed form

.. math::

    \\frac{\\partial \\ell}{\\partial x_{ij}}
    = \\frac{1}{\\sqrt{\\sigma_j^2 + \\epsilon}}
      \\left( g_{ij} - \\frac1m \\sum_k g_{kj}
        - \\hat{x}_{ij}\\,\\frac1m \\sum_k g_{kj}\\hat{x}_{kj} \\right),
    \\quad g_{ij} = \\gamma_j\\, \\frac{\\partial\\ell}{\\partial y_{ij}} .

An :func:`batch_norm_inference` helper applies the deterministic affine map used
at test time from frozen population statistics.
"""
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

__all__ = [
    "BatchNormCache",
    "batch_norm_forward",
    "batch_norm_backward",
    "batch_norm_inference",
]

Matrix = Sequence[Sequence[float]]
Vector = Sequence[float]


@dataclass(frozen=True)
class BatchNormCache:
    """Intermediate quantities saved by the forward pass for the backward pass.

    Attributes
    ----------
    x_hat:
        Standardized activations, shape ``(m, d)``.
    inv_std:
        Per-feature ``1 / sqrt(var + eps)``, shape ``(d,)``.
    gamma:
        The scale vector used in the forward pass, shape ``(d,)``.
    mean:
        Per-feature batch means, shape ``(d,)``.
    var:
        Per-feature batch variances (population, ``ddof=0``), shape ``(d,)``.
    """

    x_hat: NDArray[np.float64]
    inv_std: NDArray[np.float64]
    gamma: NDArray[np.float64]
    mean: NDArray[np.float64]
    var: NDArray[np.float64]


def _as_matrix(X: Matrix, name: str = "X") -> NDArray[np.float64]:
    arr = np.asarray(X, dtype=np.float64)
    if arr.ndim != 2:
        raise ValueError(f"{name} must be a 2-D matrix, got array with {arr.ndim} dimension(s)")
    if arr.shape[0] < 1:
        raise ValueError(f"{name} must have at least one row")
    if arr.shape[1] < 1:
        raise ValueError(f"{name} must have at least one feature")
    if not np.all(np.isfinite(arr)):
        raise ValueError(f"{name} contains non-finite values (nan or inf)")
    return arr


def _as_vector(v: Vector, d: int, name: str) -> NDArray[np.float64]:
    arr = np.asarray(v, dtype=np.float64)
    if arr.ndim != 1:
        raise ValueError(f"{name} must be a 1-D vector, got array with {arr.ndim} dimension(s)")
    if arr.shape[0] != d:
        raise ValueError(f"{name} has length {arr.shape[0]} but X has {d} features")
    if not np.all(np.isfinite(arr)):
        raise ValueError(f"{name} contains non-finite values (nan or inf)")
    return arr


def batch_norm_forward(
    X: Matrix,
    gamma: Vector,
    beta: Vector,
    *,
    eps: float = 1e-5,
) -> tuple[NDArray[np.float64], BatchNormCache]:
    """Training-time Batch Normalization forward pass.

    Parameters
    ----------
    X:
        Mini-batch of shape ``(m, d)`` with ``m >= 1`` examples and ``d >= 1``
        features. Must contain only finite values.
    gamma, beta:
        Learnable per-feature scale and shift, each of length ``d``.
    eps:
        Positive constant added to the variance for numerical stability.

    Returns
    -------
    (y, cache)
        ``y`` is the normalized, scaled and shifted output of shape ``(m, d)``.
        ``cache`` holds the quantities needed by :func:`batch_norm_backward`.

    Examples
    --------
    >>> y, _ = batch_norm_forward([[1.0], [3.0]], [1.0], [0.0])
    >>> [round(float(v), 6) for v in y.ravel()]
    [-0.999995, 0.999995]
    """
    if not (eps > 0.0):
        raise ValueError(f"eps must be positive, got {eps}")
    arr = _as_matrix(X)
    _m, d = arr.shape
    g = _as_vector(gamma, d, "gamma")
    b = _as_vector(beta, d, "beta")

    mean = arr.mean(axis=0)
    var = arr.var(axis=0)  # population variance (ddof=0)
    inv_std = 1.0 / np.sqrt(var + eps)
    x_hat = (arr - mean) * inv_std
    y = g * x_hat + b

    cache = BatchNormCache(x_hat=x_hat, inv_std=inv_std, gamma=g, mean=mean, var=var)
    return y, cache


def batch_norm_backward(
    dy: Matrix,
    cache: BatchNormCache,
) -> tuple[NDArray[np.float64], NDArray[np.float64], NDArray[np.float64]]:
    """Backward pass: gradients with respect to ``X``, ``gamma`` and ``beta``.

    Parameters
    ----------
    dy:
        Upstream gradient ``dL/dy`` of shape ``(m, d)`` matching the forward output.
    cache:
        The :class:`BatchNormCache` returned by :func:`batch_norm_forward`.

    Returns
    -------
    (dx, dgamma, dbeta)
        ``dx`` has shape ``(m, d)``; ``dgamma`` and ``dbeta`` have shape ``(d,)``.
    """
    dY = _as_matrix(dy, "dy")
    m, d = dY.shape
    if cache.x_hat.shape != dY.shape:
        raise ValueError(
            f"dy has shape {dY.shape} but cache was built for {cache.x_hat.shape}"
        )

    dgamma = np.sum(dY * cache.x_hat, axis=0)
    dbeta = np.sum(dY, axis=0)

    g = dY * cache.gamma  # dL/dx_hat
    dx = cache.inv_std * (
        g - g.mean(axis=0) - cache.x_hat * (g * cache.x_hat).mean(axis=0)
    )
    return dx, dgamma, dbeta


def batch_norm_inference(
    X: Matrix,
    gamma: Vector,
    beta: Vector,
    running_mean: Vector,
    running_var: Vector,
    *,
    eps: float = 1e-5,
) -> NDArray[np.float64]:
    """Inference-time Batch Normalization using frozen population statistics.

    Applies the deterministic affine map
    ``y = gamma * (x - running_mean) / sqrt(running_var + eps) + beta`` per feature.
    Each row is normalized independently of the others, so predictions do not
    depend on which examples are batched together.
    """
    if not (eps > 0.0):
        raise ValueError(f"eps must be positive, got {eps}")
    arr = _as_matrix(X)
    _m, d = arr.shape
    g = _as_vector(gamma, d, "gamma")
    b = _as_vector(beta, d, "beta")
    rm = _as_vector(running_mean, d, "running_mean")
    rv = _as_vector(running_var, d, "running_var")
    if np.any(rv < 0.0):
        raise ValueError("running_var must be non-negative")

    return g * (arr - rm) / np.sqrt(rv + eps) + b
